common/lib/plugins/bluegreen/blue_green_plugin.ts

```diff
@@ -0,0 +1,236 @@
+  private static readonly SUBSCRIBED_METHODS: Set<string> = new Set([
+    // We should NOT subscribe to "forceConnect" pipeline since it's used by
+    // BG monitoring, and we don't want to intercept/block those monitoring connections.
+    "connect",
+    "query"
+  ]);
+
+  private static readonly CLOSED_METHOD_NAMES: Set<string> = new Set(["end", "abort"]);
+  protected readonly pluginService: PluginService;
+  protected readonly properties: Map<string, any>;
+  protected bgProviderSupplier: BlueGreenProviderSupplier;
+  protected bgStatus: BlueGreenStatus = null;
+
+  protected bgdId: string = null;
+  protected isIamInUse: boolean = false;
+
+  protected startTimeNano: bigint = BigInt(0);
+  protected endTimeNano: bigint = BigInt(0);
+  private static provider: Map<string, BlueGreenStatusProvider> = new Map();
+
+  constructor(pluginService: PluginService, properties: Map<string, any>, bgProviderSupplier: BlueGreenProviderSupplier = null) {
+    super();
+    if (!bgProviderSupplier) {
+      bgProviderSupplier = {
+        create: (pluginService: PluginService, props: Map<string, any>, bgdId: string): BlueGreenStatusProvider => {
+          return new BlueGreenStatusProvider(pluginService, props, bgdId);
+        }
+      };
+    }
+
+    this.properties = properties;
+    this.pluginService = pluginService;
+    this.bgProviderSupplier = bgProviderSupplier;
+    this.bgdId = WrapperProperties.BGD_ID.get(this.properties).trim().toLowerCase();
+  }
+
+  getSubscribedMethods(): Set<string> {
+    return BlueGreenPlugin.SUBSCRIBED_METHODS;
+  }
+
+  async connect(
+    hostInfo: HostInfo,
+    props: Map<string, any>,
+    isInitialConnection: boolean,
+    connectFunc: () => Promise<ClientWrapper>
+  ): Promise<ClientWrapper> {
+    this.resetRoutingTimeNano();
+
+    try {
+      this.bgStatus = this.pluginService.getStatus(BlueGreenStatus, this.bgdId);
+
+      if (!this.bgStatus) {
+        return this.regularOpenConnection(connectFunc, isInitialConnection);
+      }
+
+      if (isInitialConnection) {
+        this.isIamInUse = this.pluginService.isPluginInUse(IamAuthenticationPlugin);
+      }
+
+      const hostRole: BlueGreenRole = this.bgStatus.getRole(hostInfo);
+
+      if (!hostRole) {
+        // Connection to a host that isn't participating in BG switchover.
+        return this.regularOpenConnection(connectFunc, isInitialConnection);
+      }
+
+      let client: ClientWrapper | null = null;
+      let routing: ConnectRouting | undefined = this.bgStatus.connectRouting.filter((routing: ConnectRouting) =>
+        routing.isMatch(hostInfo, hostRole)
+      )[0];
+
+      if (!routing) {
+        return this.regularOpenConnection(connectFunc, isInitialConnection);
+      }
+
+      this.startTimeNano = getTimeInNanos();
+      while (routing && !client) {
+        client = await routing.apply(this, hostInfo, props, isInitialConnection, connectFunc, this.pluginService);
+        if (!client) {
```

Comment:
nit: you can invert this to get rid of a level of nesting
```suggestion
        if (client) {
          break
        }
```